File: 1957-delete-characters-to-make-fancy-string/1957-delete-characters-to-make-fancy-string.java
class Solution {
    public String makeFancyString(String s) {
        Stack<Character> st = new Stack<>();
        
        for(int i=0;i<s.length();i++){
            char ch = s.charAt(i);
            if(st.size() >= 2 && st.peek() == ch){
                char top1 = st.peek();st.pop();
                char top2 = st.peek();st.pop();
                boolean flag = false;
                if(top1 != ch || top2 != ch){
                    flag = true;
                }
                st.push(top2);
                st.push(top1);
                if(flag)st.push(ch);
            }
            else{
                st.push(ch);
            }
        }
        
        StringBuilder ans = new StringBuilder();
        while(!st.isEmpty()){
            ans.append(st.peek());
            st.pop();
        }
        return ans.reverse().toString();
        
    }
}

File: 1957-delete-characters-to-make-fancy-string/1957-delete-characters-to-make-fancy-string.cpp
class Solution {
public:
    string makeFancyString(string s) {
        stack<char> st;
        for(int i=0;i<s.size();i++){
            if(st.size() >= 2 && st.top() == s[i]){
                char top1 = st.top();st.pop();
                char top2 = st.top();st.pop();
                bool flag = false;
                if(top1 != s[i] || top2 != s[i]){
                    flag = true;
                }
                st.push(top2);
                st.push(top1);
                if(flag)st.push(s[i]);
            }
            else{
                st.push(s[i]);
            }
        }
        string ans;
        while(!st.empty()){
            ans += st.top();
            st.pop();
        }
        reverse(ans.begin(),ans.end());
        return ans;
    }
};